sinatra-contrib/lib/sinatra/config_file.rb

```diff
@@ -149,23 +139,32 @@ def message
 
     private
 
-    # Given a +hash+ with some application configuration it returns the
-    # settings applicable to the current environment.  Note that this can only
-    # be done when all the keys of +hash+ are environment names included in the
-    # +environments+ setting (which is an Array of Strings).  Also, the
-    # returned config is a indifferently accessible Hash, which means that you
-    # can get its values using Strings or Symbols as keys.
-    def config_for_env(hash)
-      if hash.respond_to?(:keys) && hash.keys.all? { |k| environments.include?(k.to_s) }
-        hash = hash[environment.to_s] || hash[environment.to_sym]
-      end
+    # Given a +config+ hash with some application configuration, returns the settings
+    # applicable to the current +environment+.
+    def config_for_env(config)
+      return config[environment.to_s] || {} if has_environment_keys?(config)
 
-      if hash.respond_to?(:to_hash)
-        IndifferentHash[hash.to_hash]
-      else
-        hash
+      config.each_with_object({}) do |(key, value), acc|
+        value = value[environment.to_s] if has_environment_keys?(value)
+        acc.merge!(key => with_indifferent_access(value)) unless value.nil?
       end
     end
+
+    # Returns true if supplied with a hash that has any recognized
+    # +environments+ in its root keys.
+    def has_environment_keys?(hash)
+      return false unless hash.is_a?(Hash)
+
+      hash.keys.map(&:to_s).any? { |k| environments.include?(k) }
+    end
+
+    # Returns a hash that can be accessed by both strings and symbols, when
+    # supplied with a hash with string keys.
+    def with_indifferent_access(hash)
```

Comment:
Do we need Yet Another Indifferent Hash implementation? Can't we use the one in core?

Reply:
I think this PR precedes the work to build out a full IndifferentHash implementation. I can definitely look at switching to that today.